### DISCUSSION TITLE
Unable to run METviewer container

### DISCUSSION THREAD

**Discussion by jwraby** [dtcenter/METplus#2505] [Plot Generation] [answered]
I successfully pulled the Docker container (metviewer_latest) from docker://dtcenter/metviewer and converted it to a Singularity container on the HPC. The installation appeared to be successful, but when I run the container, it gives the following error:
true: /opt/tomcat/bin/startup.sh: Permission denied
I assume that this error is occurring because the container cannot run on the HPC. Is it possible to obtain a METviewer container which will run on an HPC? Thanks.

**Comment by jprestop** [comment]
@jwraby, we're working on finding the best person on our team to help with this task.  Someone should follow up soon.  Thank you for your patience.

**Reply by jwraby** [reply]
@jprestop, Thanks for the update. Understand and there is no rush.

**Comment by TatianaBurek** [comment]
METviewer does not officially support the run of Singularity containers on HPC. We have an issue created sometime ago to implement this feature : https://github.com/dtcenter/METviewer/issues/281
However, some work was done for one of the protects that required a METviewer container running on HPC. For that, we created a custom Dockefile that builds an image that can be converted into the Singularity image for HPC:
https://github.com/dtcenter/METviewer/blob/main_v5.1/docker/Dockerfile_for_Singularity
And this is the issue that describes how to build and run it: https://github.com/dtcenter/METviewer/issues/446
You can build the container or get it from the repository

**Comment by jwraby** [comment]
Hi Tatiana. Thanks for this information. When I pulled the METviewer container from the Docker repository, was that container the one you describe which was built from the custom Dockerfile? Are you able to bypass the use of Tomcat and produce batch plots without a web server?

**Comment by TatianaBurek** [comment]
The "Singularity" docker container is located in dtcenter/nwp-container-metviewer-for-singularity repository (https://github.com/NCAR/container-dtc-nwp/issues/94#issuecomment-1241204523)
dtcenter/metviewer has a standard docker container.
We could not use the web app/Tomcat features because of the disabled ports but were able to use  batch scripts

**Comment by jwraby** [comment]
I tried to pull the container as follows, but get an error:
singularity pull docker://dtcenter/nwp-container-metviewer-for-singularity
FATAL:   While making image from oci registry: error fetching image to cache: failed to get checksum for docker://dtcenter/nwp-container-metviewer-for-singularity: Error reading manifest latest in docker.io/dtcenter/nwp-container-metviewer-for-singularity: manifest unknown: manifest unknown
Do you know what the problem is?

**Reply by jwraby** [reply]
Tatiana - I was looking at my logs and found that I had the same error when attempting to pull the latest METplus docker container. John Halley Gotway recommended that I specify the tag name of the METplus container when I execute the pull command. I guess for the METviewer container that might look like nwp-container-metviewer-for-singularity:5.1.0. Would this work for the METviewer container?

**Reply by TatianaBurek** [reply]
nwp-container was built for METviewer 4.1:

singularity build --sandbox --fix-perms --force metv4singularity docker://dtcenter/**nwp-container-metviewer-for-singularity:4.1.0**
or 
docker pull dtcenter/nwp-container-metviewer-for-singularity:4.1.0

**Reply by jwraby** [reply]
The information was very helpful. Successfully pulled and converted the container to Singularity on the HPC. I tried to translate the command you used to run the Docker container from https://github.com/dtcenter/METviewer/issues/446 into the equivalent command for Singularity: 
Docker command: docker run --name=mv1 -d -p 8080:8080 -e MYSQL_ROOT_PASSWORD=xxxxxx
Singularity command: singularity run --hostname=mv1 --env MYSQL_ROOT_PASSWORD=xxxxxx nwp-container-metviewer-for-singularity_4.1.0.sif

I am not sure that my translation is completely correct. For Singularity I chose not to run in the background (-d). I also chose not to specify the container port (8080) because I cannot run Tomcat. I assume that "container name" for Docker is the same as "container hostname" in Singularity. For Singularity, I used "--env string" to set the environment variable. 

I ran the above Singularity command and received the following error:
Running mysql_install_db ...
mkdir: cannot create directory ‘/var/lib/mysql/mysql’: Read-only file system
Fatal error Can't create database directory '/var/lib/mysql/mysql'

What is your diagnosis for this error? Is there a possible resolution to overcome this error?

**Comment by TatianaBurek** [comment, marked as answer]
https://github.com/NCAR/container-dtc-nwp/issues/94 contains the instructions on how to run the singularity container:
https://github.com/NCAR/container-dtc-nwp/issues/94#issuecomment-1241204523
it also has a link to the online tutorial :
https://dtcenter.org/nwp-containers-online-tutorial/software-containers/metviewer-database-display

**Reply by jwraby** [reply]
Thanks. I followed the comment section at the above link and I think I successfully started the singularity instance as follows:
cd /p/home/jraby/MET/METviewer_NWP_singularity
singularity instance start --writable metv4singularity metv
INFO:    instance started successfully
However, on executing the next command from the comment section I get an error as follows:
singularity exec --writable instance://metv bash init_singularity.sh
/usr/bin/bash: init_singularity.sh: No such file or directory

I navigated to: /p/home/jraby/MET/METviewer_NWP_singularity/metv4singularity/usr/bin and looked for a subdirectory named "bash" but I can't find it. What is your diagnosis for this error?

**Reply by jwraby** [reply]
I found the script which I think it is trying to run at https://github.com/dtcenter/METviewer/issues/446. The script is init_singularity.sh and there are links to download a zip file containing the script, but the download doesn't work. This site can’t be reachedobjects.githubusercontent.com refused to connect. I tried using the other link, but it also has an error. 404 - page not found
Cannot find a valid ref in issue_446_container_singularity/docker/init_singularity.sh. I will continue poking around on GitHub to see if I can find the script.

**Reply by jwraby** [reply]
I copied the text from the script and pasted it into a text doc named "init_singularity.sh". I copied this file to the home directory on the HPC  /p/home/jraby/MET/METviewer_NWP_singularity. I re-ran the command at this location as follows:
singularity exec --writable instance://metv bash init_singularity.sh
It appears to be running, but it was aborted. See following dialog:
Running mysql_install_db ...
Installing MariaDB/MySQL system tables in '/var/lib/mysql' ...
240227 22:27:51 [Note] /usr/libexec/mysqld (mysqld 5.5.68-MariaDB) starting as process 62 ...
ERROR: 1049  Unknown database 'mysql'
240227 22:27:51 [ERROR] Aborting

Discovered a possible problem with setting the env variables in the script. Modified the statements in the script to add "export" as follows:
export DATADIR="/var/lib/mysql"
export MYSQL_ROOT_PASSWORD='mvuser'

Re-ran the command again:
singularity exec --writable instance://metv bash init_singularity.sh
See the following dialog:
Running mysql_install_db ...
Installing MariaDB/MySQL system tables in '/var/lib/mysql' ...
240227 22:45:45 [Note] /usr/libexec/mysqld (mysqld 5.5.68-MariaDB) starting as process 263 ...
ERROR: 1  Can't create/write to file '/p/work1/tmp/jraby/#sql_107_0.MAI' (Errcode: 2)
240227 22:45:45 [ERROR] Aborting

240227 22:45:45 [Note] /usr/libexec/mysqld: Shutdown complete


Installation of system tables failed!  Examine the logs in
/var/lib/mysql for more information.

The problem could be conflicting information in an external
my.cnf files. You can ignore these by doing:

    shell> /usr/bin/mysql_install_db --defaults-file=~/.my.cnf

You can also try to start the mysqld daemon with:

    shell> /usr/libexec/mysqld --skip-grant-tables --general-log &

and use the command line tool /usr/bin/mysql
to connect to the mysql database and look at the grant tables:

    shell> /usr/bin/mysql -u root mysql
    mysql> show tables;

Try 'mysqld --help' if you have problems with paths.  Using
--general-log gives you a log in /var/lib/mysql that may be helpful.

The latest information about mysql_install_db is available at
https://mariadb.com/kb/en/installing-system-tables-mysql_install_db
You can find the latest source at https://downloads.mariadb.org and
the maria-discuss email list at https://launchpad.net/~maria-discuss

Please check all of the above before submitting a bug report
at http://mariadb.org/jira

Finished mysql_install_db
UMASK must be a 3-digit mode with an additional leading 0 to indicate octal.
The first digit will be corrected to 6, the others may be 0, 2, 4, or 6.
UMASK corrected from 002 to 0640 ...
Tomcat started.
(base) [jraby@gaffney03 METviewer_NWP_singularity]$ 240227 22:45:46 mysqld_safe Logging to '/var/log/mariadb/mariadb.log'.
240227 22:45:46 mysqld_safe Starting mysqld daemon with databases from /var/lib/mysql

It seems to be hung at this point with no more dialog being written. I'm not sure how to proceed from here.

**Reply by jwraby** [reply]
I hit "enter" to see what happens and the normal prompt returned. 
The next command is:
singularity shell instance://metv

I ran this and it started the container with the expected Singularity> prompt.

At this point the container seems to be running. I assume that everything is ready for me to create a METViewer database and then populate it with schema. I have instructions which I have used in the past for METviewer 4.1.0 installed on a local machine, but for that I ran database commands as follows:
i.	Navigate to /opt/METviewer_py3.8.14/METviewer/bin
ii.	mysql -h 127.0.0.1 -u xxxxxxxxx -p -e 'create database mv_test1'
iii.	mysql -h 127.0.0.1 -u xxxxxxxxx -p mv_test1</opt/METviewer_py3.8.14/METviewer/sql/mv_mysql.sql

I plan to copy the mv_load.sh script, the load_spec_file and the MET STAT files which I used for testing on the local machine to the HPC. How do I execute the commands to create and populate the database and then where do I put the mv_load.sh script, the load_spec_file and the MET STAT files so the container will work to generate the batch plots?

**Reply by TatianaBurek** [reply]
Looking on [metv_load_singularity.ksh](https://github.com/NCAR/container-dtc-nwp/blob/main/components/scripts/common/metv_load_singularity.ksh) I see the commands that are used to create and load a database:
```
mysql -hlocalhost -uroot -pmvuser -e"drop database <database_name>;"
mysql -hlocalhost -uroot -pmvuser -e"create database <database_name>;"
mysql -hlocalhos -uroot -pmvuser <database_name> < /METviewer/sql/mv_mysql.sql
```

**Reply by jwraby** [reply]
Thanks. These commands look familiar. At what point do I execute the commands to create and load a database?
Before or after executing the following?
singularity instance start --writable metv4singularity metv
singularity exec --writable instance://metv bash init_singularity.sh
singularity shell instance://metv
I'm not sure what state the container must be in in order to create and load the database.

**Reply by TatianaBurek** [reply]
Database creating and loading should happen after 
`singularity shell instance://metv`
I would suggest to check if the database is properly working by logging in/logging out before creating a database:
`mysql -hlocalhost -uroot -pmvuser`

**Reply by jwraby** [reply]
Thanks for the suggestion. I ran the following commands to set-up and run the container before executing the mysql command:
pwd
/p/home/jraby/MET/METviewer_NWP_singularity
ll
total 8
-rwx------.  1 jraby 1487I100  719 Feb 27 22:45 init_singularity.sh
drwxr-x---. 21 jraby 1487I100 4096 Feb 27 17:17 metv4singularity
singularity instance start --writable metv4singularity metv
FATAL:   instance metv already exists
singularity exec --writable instance://metv bash init_singularity.sh
Running mysql_install_db ...
Installing MariaDB/MySQL system tables in '/var/lib/mysql' ...
240305 17:53:54 [Note] /usr/libexec/mysqld (mysqld 5.5.68-MariaDB) starting as process 898 ...
ERROR: 1  Can't create/write to file '/p/work1/tmp/jraby/#sql_382_0.MAI' (Errcode: 2)
240305 17:53:54 [ERROR] Aborting

240305 17:53:54 [Note] /usr/libexec/mysqld: Shutdown complete


Installation of system tables failed!  Examine the logs in
/var/lib/mysql for more information.

The problem could be conflicting information in an external
my.cnf files. You can ignore these by doing:

    shell> /usr/bin/mysql_install_db --defaults-file=~/.my.cnf

You can also try to start the mysqld daemon with:

    shell> /usr/libexec/mysqld --skip-grant-tables --general-log &

and use the command line tool /usr/bin/mysql
to connect to the mysql database and look at the grant tables:

    shell> /usr/bin/mysql -u root mysql
    mysql> show tables;

Try 'mysqld --help' if you have problems with paths.  Using
--general-log gives you a log in /var/lib/mysql that may be helpful.

The latest information about mysql_install_db is available at
https://mariadb.com/kb/en/installing-system-tables-mysql_install_db
You can find the latest source at https://downloads.mariadb.org and
the maria-discuss email list at https://launchpad.net/~maria-discuss

Please check all of the above before submitting a bug report
at http://mariadb.org/jira

Finished mysql_install_db
UMASK must be a 3-digit mode with an additional leading 0 to indicate octal.
The first digit will be corrected to 6, the others may be 0, 2, 4, or 6.
UMASK corrected from 002 to 0640 ...
Tomcat started.
(base) [jraby@gaffney03 METviewer_NWP_singularity]$ 240305 17:53:54 mysqld_safe Logging to '/var/log/mariadb/mariadb.log'.
240305 17:53:54 mysqld_safe Starting mysqld daemon with databases from /var/lib/mysql
singularity shell instance://metv
Singularity> mysql -h localhost -uroot -pmvuser
ERROR 2002 (HY000): Can't connect to local MySQL server through socket '/var/lib/mysql/mysql.sock' (2)

**Reply by TatianaBurek** [reply]
Looks like you have a problem creating tmp files for the database. I assume that you don't have 'write' permissins in '/p/work1/tmp/' directory or this directory is full.
I remember that we had similar issues. Try to use [this](https://github.com/dtcenter/METviewer/blob/main_v5.1/docker/init_singularity.sh) init_singularity.sh script instead. It assigns the TMPDIR location to your home directory.

**Comment by jwraby** [comment]
Thanks for looking at this error. I checked the status of '/p/work1/tmp/' and it is a directory containing what appear to be users accounts on the HPC, including mine 'jraby'. I tried to create a file in '/p/work1/tmp/', but I don't have write permission which agrees with your assumption. I went into 'jraby" and I do have write permissions there. The vol of jraby is 40K. 
I will try using the init_singularity.sh script you provided to see what happens.

**Reply by jwraby** [reply]
Good news! I used the new script you provided and the results appear to be successful. See the following dialog which was written after I executed the script:
singularity exec --writable instance://metv bash init_singularity.sh
Directory /p/home/jraby/tmp is created.
Running mysql_install_db ..........
Installing MariaDB/MySQL system tables in '/var/lib/mysql' ...
240307 15:58:04 [Note] /usr/libexec/mysqld (mysqld 5.5.68-MariaDB) starting as process 44 ...
OK
Filling help tables...
240307 15:58:04 [Note] /usr/libexec/mysqld (mysqld 5.5.68-MariaDB) starting as process 53 ...
OK

To start mysqld at boot time you have to copy
support-files/mysql.server to the right place for your system

PLEASE REMEMBER TO SET A PASSWORD FOR THE MariaDB root USER !
To do so, start the server, then issue the following commands:

'/usr/bin/mysqladmin' -u root password 'new-password'
'/usr/bin/mysqladmin' -u root -h gaffney01 password 'new-password'

Alternatively you can run:
'/usr/bin/mysql_secure_installation'

which will also give you the option of removing the test
databases and anonymous user created by default.  This is
strongly recommended for production servers.

See the MariaDB Knowledgebase at http://mariadb.com/kb or the
MySQL manual for more instructions.

You can start the MariaDB daemon with:
cd '/usr' ; /usr/bin/mysqld_safe --datadir='/var/lib/mysql'

You can test the MariaDB daemon with mysql-test-run.pl
cd '/usr/mysql-test' ; perl mysql-test-run.pl

Please report any problems at http://mariadb.org/jira

The latest information about MariaDB is available at http://mariadb.org/.
You can find additional information about the MySQL part at:
http://dev.mysql.com
Consider joining MariaDB's strong and vibrant community:
https://mariadb.org/get-involved/

Finished mysql_install_db ..........
Starting MariaDB  ..........
Starting Tomcat  ..........
UMASK must be a 3-digit mode with an additional leading 0 to indicate octal.
The first digit will be corrected to 6, the others may be 0, 2, 4, or 6.
(base) [jraby@gaffney01 METviewer_NWP_singularity]$ UMASK corrected from 002 to 0640 ...
Tomcat started.
240307 15:58:04 mysqld_safe Logging to '/var/log/mariadb/mariadb.log'.
240307 15:58:05 mysqld_safe Starting mysqld daemon with databases from /var/lib/mysql


A this point, the command prompt has not returned, but it appears the process is running. Hit 'enter' to restore the command prompt.

Started the container:
singularity shell instance://metv
Singularity>
mysql -hlocalhost -uroot -pmvuser
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 2
Server version: 5.5.68-MariaDB MariaDB Server

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [(none)]>

At this point, the database appears to be running OK and I was able to explore the contents as follows:

MariaDB [(none)]> show databases;
+--------------------+
| Database           |
+--------------------+
| information_schema |
| mysql              |
| performance_schema |
+--------------------+
3 rows in set (0.00 sec)

MariaDB [(none)]> show tables;
ERROR 1046 (3D000): No database selected

MariaDB [(none)]> use information_schema;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed

MariaDB [information_schema]> quit;
Bye
Singularity>

Assuming everything is installed and ready for use, what are my options for interacting with METviewer to generate plots?

**Reply by jwraby** [reply]
Update at the end of the day. I was able to successfully create a database, add the schema, and then load the database with STAT files. Can I use the web GUI to generate plots or do I need to use the batch commands? I haven't used batch commands before. Do you have some examples?
Thanks.

**Reply by jwraby** [reply]
I found good documentation at https://metviewer.readthedocs.io/en/latest/Users_Guide/batch.html, Will this work for the container version I have running?

I have some plot_spec files I generated when I produced some plots using the web GUI on my local machine (METviewer V4.1.0). Could I use these for testing batch plots on the HPC?

Thanks.

**Reply by jwraby** [reply]
Having difficulty locating <folders> specs in the plot_spec file. 
For <r_tmpl>/opt/tomcat/webapps/R_tmpl</r_tmpl>
and <r_work>/opt/tomcat/webapps/R_work</r_work> I was able to find the equivalent paths as follows: 
/p/home/jraby/MET/METviewer_NWP_singularity/metv4singularity/METviewer
and at: /p/home/jraby/MET/METviewer_NWP_singularity/metv4singularity/METviewer/dist
I can't find the equivalents for:
<plots>/opt/tomcat/webapps/metviewer_output/plots</plots>
<data>/opt/tomcat/webapps/metviewer_output/data</data>
<scripts>/opt/tomcat/webapps/metviewer_output/scripts</scripts>
Will continue searching.

**Reply by jwraby** [reply]
Continued searching in the container and found likely directories and files which appeared to satisfy the spec. I created new directories in the container for the following:
/opt/tomcat/webapps/metviewer_output/plots
/opt/tomcat/webapps/metviewer_output/data
/opt/tomcat/webapps/metviewer_output/scripts

Edited my old plot_spec file as needed for the container and database installed on the HPC and then successfully ran the mv_batch.sh script. It generated two plots using the series_plot.R_tmpl script (.html and .png). The output plot matched the one I had generated in JAN 2023 using METviewer V4.1.0 on my local machine.

From the plot dialog written in the terminal, it appears that a Python script was run to produce the plot.

Is it possible to use the web GUI on the HPC to produce plots?

I now have METviewer V4.1.0 installed and running on the HPC. Your expertise and patience for this discussion are very much appreciated!

**Reply by TatianaBurek** [reply]
To run GUI you need to have a port 8080 to be open on the system. Our servers have this and other possible ports closed for the security reasons. So, no, it was impossible in our situation to use GUI.
But wen we used the  same Singularity container on AWS where we dictated the system requirements and permissions we were able to use the web GUI

**Reply by jwraby** [reply]
Is the port 8080 part of the HPC system? I'm not sure I understand how a port works. It seems like another way to communicate with the HPC. For example, I use PuTTY to open a terminal connection to the HPC and that uses port 22 as one of the settings.  So, the HPC has a port 22 open to allow terminal sessions. If I am understanding this correctly, having port 8080 open, means that I can interact with the HPC using a web browser. Am I understanding this  correctly? Thanks.

**Reply by jwraby** [reply]
I asked the HPC system admin if port 8080 was open. Their response was "8080 is not open on the firewall" My take is that this means that I will not be able to use the METviewer GUI. My approach will be to learn how to produce plots using the batch method. Can the batch method produce all the same types of plots as the GUI?

**Reply by TatianaBurek** [reply]
We have the same situation with port 8080 being not available.
And yes, you can create the same plots with batch as with GUI

**Reply by jwraby** [reply]
Thanks again for your valuable assistance with getting the METview container working on the HPC and for answering all my questions. I've learned a lot from this! I may have more METviewer questions in the future, but for now I think this discussion has achieved all the results I was hoping for!

**Comment by georgemccabe** [comment]
@jvigh, I wanted to bring your attention to this discussion in case it contains useful information to help with the issues you were experiencing.
